File: solutions/ruby/dnd-character/1/dnd_character.rb
=begin
Write your code for the 'D&D Character' exercise in this file. Make the tests in
`dnd_character_test.rb` pass.

To get started with TDD, see the `README.md` file in your
`ruby/dnd-character` directory.
=end

class DndCharacter
  
  def self.modifier(constitution)
    ((constitution - 10) / 2).floor
  end

  attr_reader :strength, :dexterity, :constitution, :intelligence, :wisdom, :charisma, :hitpoints

  def initialize
    @strength     = Random.random_number(3..18)
    @dexterity    = Random.random_number(3..18)
    @constitution = Random.random_number(3..18)
    @intelligence = Random.random_number(3..18)
    @wisdom       = Random.random_number(3..18)
    @charisma     = Random.random_number(3..18)
    @hitpoints    = 10 + DndCharacter.modifier(@constitution)
  end
  
end

File: solutions/lua/dnd-character/1/dnd-character.lua
local Character = {}
Character.__index = Character

local function ability(scores)
  local score = 0
  local lowest = 0
  for i = 1, 4 do
    local roll = math.floor(math.random() * 6 + 1)
    score = score + roll
    if lowest == 0 or lowest > roll then
      lowest = roll
    end
  end
  return score - lowest
end

local function modifier(constitution)
  return math.floor((constitution - 10) / 2)
end

function Character:new(name)
  local new_character = {
    name = name,
    strength = ability(),
    dexterity = ability(),
    constitution = ability(),
    intelligence = ability(),
    wisdom = ability(),
    charisma = ability(),
  }

  new_character['hitpoints'] = 10 + modifier(new_character.constitution)
  
  return setmetatable(new_character, self)
end

return {
  Character = Character,
  ability = ability,
  modifier = modifier
}
